url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/de/odysseus/el/tree/impl/ast/AstFunction.java

```diff
@@ -0,0 +1,166 @@
+		this.name = name;
+		this.index = index;
+		this.params = params;
+		this.varargs = varargs;
+	}
+
+	/**
+	 * Invoke method.
+	 * @param bindings
+	 * @param context
+	 * @param base
+	 * @param method
+	 * @return method result
+	 * @throws InvocationTargetException
+	 * @throws IllegalAccessException
+	 */
+	protected Object invoke(Bindings bindings, ELContext context, Object base, Method method)
+		throws InvocationTargetException, IllegalAccessException {
+		Class<?>[] types = method.getParameterTypes();
+		Object[] params = null;
+		if (types.length > 0) {
+			params = new Object[types.length];
+			if (varargs && method.isVarArgs()) {
+				for (int i = 0; i < params.length - 1; i++) {
+					Object param = getParam(i).eval(bindings, context);
+					if (param != null || types[i].isPrimitive()) {
+						params[i] = bindings.convert(param, types[i]);
+					}
+				}
+				int varargIndex = types.length - 1;
+				Class<?> varargType = types[varargIndex].getComponentType();
+				int length = getParamCount() - varargIndex;
+				Object array = null;
+				if (length == 1) { // special: eventually use argument as is
+					Object param = getParam(varargIndex).eval(bindings, context);
+					if (param != null && param.getClass().isArray()) {
+						if (types[varargIndex].isInstance(param)) {
+							array = param;
+						} else { // coerce array elements
+							length = Array.getLength(param);
+							array = Array.newInstance(varargType, length);
+							for (int i = 0; i < length; i++) {
+								Object elem = Array.get(param, i);
+								if (elem != null || varargType.isPrimitive()) {
+									Array.set(array, i, bindings.convert(elem, varargType));
+								}
+							}
+						}
+					} else { // single element array
+						array = Array.newInstance(varargType, 1);
+						if (param != null || varargType.isPrimitive()) {
+							Array.set(array, 0, bindings.convert(param, varargType));
+						}
+					}
+				} else {
+					array = Array.newInstance(varargType, length);
+					for (int i = 0; i < length; i++) {
+						Object param = getParam(varargIndex + i).eval(bindings, context);
+						if (param != null || varargType.isPrimitive()) {
+							Array.set(array, i, bindings.convert(param, varargType));
+						}
+					}
+				}
+				params[varargIndex] = array;
+			} else {
+				for (int i = 0; i < params.length; i++) {
+					Object param = getParam(i).eval(bindings, context);
+					if (param != null || types[i].isPrimitive()) {
+						params[i] = bindings.convert(param, types[i]);
+					}
+				}
+			}
+		}
+		return method.invoke(base, params);
+	}
+
+	@Override 
+	public Object eval(Bindings bindings, ELContext context) {
+		Method method = bindings.getFunction(index);
+		try {
+			return invoke(bindings, context, null, method);
+		} catch (IllegalAccessException e) {
+			throw new ELException(LocalMessages.get("error.function.access", name), e);
+		} catch (InvocationTargetException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)